src/oc_erchef/apps/chef_objects/src/chef_cookbook_version.erl

```diff
@@ -589,6 +672,84 @@ list_query(_ObjectRec) ->
 bulk_get_query(_ObjectRec) ->
     error(not_implemented).
 
+%% for v2 metadata, the "segment" is encoded in to the name of file
+%% so for data that is v0 on disk we have to prepend the segment to the name
+add_segment_to_filename(Segment, File) ->
+    Fn = ej:get({<<"name">>}, File),
+    Fn1 = iolist_to_binary([Segment, "/", Fn]),
+    ej:set({<<"name">>}, File, Fn1).
+
+remove_segment_from_filename(File) ->
+    [Segment | Name ] = get_segment_from_record(File),
+    case Name of
+        % if Name is nil, then we have a root file (like metadata.rb) - so we'll set the name to the segment, and the segment to <<"root_files">>
+        [] ->
+            { <<"root_files">>, ej:set({<<"name">>}, File, Segment)};
+        _ ->
+            Record = ej:set({<<"name">>}, File, lists:last(Name)),
+            { Segment, Record }
+    end.
+
+populate_all_files(Segment, Data, Metadata) ->
+    lists:foldl(fun(File, CB) ->
+                        File1 = case Segment of
+                                    <<"root_files">> ->
+                                        File;
+                                    _ ->
+                                        add_segment_to_filename(Segment, File)
+                                end,
+                        ej:set_p({<<"all_files">>, new}, CB, File1)
+                end,
+                Metadata,
+                Data).
+
+%% A v2 cookbook version contains only the "all_files" key, which is a list of all the file parts
+%% { "all_files": [ { "name": "recipes/default.rb", "path": "recipes/default.rb", … } ] }
+%% We have to transform that into segments (?COOKBOOK_SEGMENTS) containing a list of file parts for that segment
+%% { "recipes": [ { "name": "default.rb", path: "recipes/default.rb", … } ] }
+populate_segments(Data, Metadata) ->
+    CB1 = lists:foldl(fun(Segment, CB) ->
+                              ej:set({Segment}, CB, [])
+                      end,
+                      Metadata,
+                      ?COOKBOOK_SEGMENTS),
+
+    lists:foldl(fun(File, CB) ->
+                        { Segment, Record } = remove_segment_from_filename(File),
+                        case lists:member(Segment, ?COOKBOOK_SEGMENTS) of
+                            true -> ej:set_p({Segment, new}, CB, Record);
```

Comment:
I tend to do much of my building of JSON structures outside of EJSON, then insert in the end. IMHO it can lead to clearer code. In particular, the ej:set_p pattern with new caught me by surprise, (it's not used anywhere else in chef-server or a1 that I can see.
I might have done this by folding over a map of segments to arrays, inserting into that, and then for each recognized COOKBOOK_SEGMENT extracting the built array from the map and ej:set-ting the field.
That's a matter of taste though.
As a secondary thing, I do like splitting things into smaller functions when possible; mostly because redbug doesn't really let you trace an anonymous function easily.

Reply:
So you'd break out the function the second map takes (assuming the structure remains the same) into a real function?